proposals/2022-09-14-kubernetes-versioning.md

```diff
@@ -0,0 +1,208 @@
+2. The following shows the containerd config.toml that maps the runtime type to the shim binary from step 1.
+      ```toml
+        [plugins.cri.containerd.runtimes.spin-v1]
+          runtime_type = "io.containerd.spin-v1.2.3.v1"
+        [plugins.cri.containerd.runtimes.spin-v1_2]
+          runtime_type = "io.containerd.spin-v1.2.3.v1"
+        [plugins.cri.containerd.runtimes.spin-v1_2_3]
+          runtime_type = "io.containerd.spin-v1.2.3.v1"
+      ```
+3. A runtime class that contains a handler that matches the "spin" config runtime name from step 2.
+      ```yaml
+      ---
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-spin-v1.2.3
+      handler: spin-v1_2_3
+      scheduling:
+        nodeSelector:
+          wasmtime-spin-v1.2.3: "true"
+      ---
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-spin-v1.2
+      handler: spin-v1_2
+      scheduling:
+        nodeSelector:
+          wasmtime-spin-v1.2: "true"
+      ---
+      apiVersion: node.k8s.io/v1
+      kind: RuntimeClass
+      metadata:
+        name: wasmtime-spin-v1
+      handler: spin-v1
+      scheduling:
+        nodeSelector:
+          wasmtime-spin-v1: "true"
+      ```
+
+    **NOTE: node labels would be expected for both `wasmtime-spin-v1: "true"` and `wasmtime-spin-v1.2.3: "true"` on nodes that have the spin shim v1.2.3 installed.**
+
+4. A Wasm workload with the specified RuntimeClass name matching the "wasmtime-spin-v1" runtime class from step 3. The "wasmtime-spin-v1" runtimeClassName would map to the spin-v1.2.3 handler.
+      ```yaml
+      apiVersion: apps/v1
+      kind: Deployment
+      metadata:
+        name: wasm-spin
+      spec:
+        replicas: 1
+        selector:
+          matchLabels:
+            app: wasm-spin
+        template:
+          metadata:
+            labels:
+              app: wasm-spin
+          spec:
+            runtimeClassName: wasmtime-spin-v1
+            containers:
+            - name: spin-hello
+              image: ghcr.io/deislabs/containerd-wasm-shims/examples/spin-rust-hello:latest
+              command: ["/"]
+      ```
+
+### <a name='Introducinganewversionofashim'></a>Introducing a new version of a shim
+Now that we have described the versioning scheme, let's describe what happens when we introduce a new node that contains a new version of a shim. In this example, we will introduce v1.2.4 of the Spin shim.
+
+For this example, let's imagine we have a cluster with 1 node configured as described in [Versioning of both shim binary and RuntimeClass name](#versioning-of-both-shim-binary-and-runtimeclass-name). We introduce a new node to the cluster with `/bin/containerd-shim-spin-v1.2.4-v1` installed and the following containerd config section.
+
+```toml
+[plugins.cri.containerd.runtimes.spin-v1]
+runtime_type = "io.containerd.spin-v1.2.4.v1"
```

Comment:
Having to include the version in the file name itself isn't super resilient. I assume there isn't a standard runtime interface imposed upon containerd shim runtime that would allow this metadata to be introspected more deterministically? For example, of a containerd shim runtime in order to be a proper containerd shim runtime must expose a `--version` runtime context (using a standard unix CLI arg as a metaphor here) which returned a semantic version string of its instrinsic version metadata as computed at build time, then we could rely upon that instead of a filename (which could easily and dangerously be sort of overloaded when folks are trying to "force" a version-required pod spec to run on this shim.

Reply:
An alternative approach (assuming that the shim runtime itself doesn't carry integral version information) is to define a distinct, sibling property (e.g., `version`) that expreses this configuration.

```
[plugins.cri.containerd.runtimes.spin-v1]
  runtime_type = "io.containerd.spin"
  version: "v1.2.4"
```

Can I assume even if the above is more preferable that it wasn't considered because it would require evolving the Kubernetes `RuntimeClass` spec (a non-trivial prerequisite)? E.g.:

```
apiVersion: node.k8s.io/v1
kind: RuntimeClass
metadata:
  name: wasmtime-spin
  version: v1.2.4
handler: spin-v1_2_4
scheduling:
  nodeSelector:
    wasmtime-spin-v1.2.4: "true"
```

Now that I am napkin'ing this out, I can see that the `nodeSelector` itself has to properly contain the version-specific information in order to ensure that the k8s scheduler sends the pod to the correct node. (Although this does bring up a sort of gap: a node should be able to declare `RuntimeClass` support as a 1st class configuration so that `RuntimeClass` can be useful context not just for the last-mile CRI implementation details, but for the scheduler as well so that the "first-mile" story is properly fulfilled.)

Reply:
I don't believe version is available in the containerd configuration, and extending RuntimeClass is something that would take far too long to gain acceptance, if possible.

In this proposal I wanted to be as practical as possible given the constrains as they are.